core/src/main/java/org/elasticsearch/env/Environment.java

```diff
@@ -108,28 +109,32 @@
     }
 
     public Environment(Settings settings) {
+        this(PathUtils.getDefaultFileSystem(), settings);
+    }
+
+    public Environment(FileSystem fs, Settings settings) {
```

Comment:
Unfortunately this does not immediately work because it overrides the filesystem for everything. In particular, 

```
private final Path tmpFile = PathUtils.get(System.getProperty("java.io.tmpdir"));
```

in `Environment` will be resolved against the installed filesystem. Since `java.io.tmpdir` will be a path that is relevant to the default filesystem for the system that the test is executing on, it will for example cause tests that are running on a filesystem that handles paths differently to fail (e.g., running on OS X against a filesystem that is configured like Windows). A possibility exists to be evil (these tests are already evil) and change the system property for `java.io.tmpdir` before the tests start (and reset it afterwards) but that's tending towards hacky.

```diff
diff --git a/core/src/main/java/org/elasticsearch/env/Environment.java b/core/src/main/java/org/elasticsearch/env/Environment.java
index ca7363a..d89d51a 100644
--- a/core/src/main/java/org/elasticsearch/env/Environment.java
+++ b/core/src/main/java/org/elasticsearch/env/Environment.java
@@ -31,7 +31,6 @@ import java.net.MalformedURLException;
 import java.net.URISyntaxException;
 import java.net.URL;
 import java.nio.file.FileStore;
-import java.nio.file.FileSystem;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.ArrayList;
@@ -109,32 +108,28 @@ public class Environment {
     }
 
     public Environment(Settings settings) {
-        this(PathUtils.getDefaultFileSystem(), settings);
-    }
-
-    public Environment(FileSystem fs, Settings settings) {
         this.settings = settings;
         final Path homeFile;
         if (PATH_HOME_SETTING.exists(settings)) {
-            homeFile = PathUtils.get(fs, PATH_HOME_SETTING.get(settings));
+            homeFile = PathUtils.get(PATH_HOME_SETTING.get(settings));
         } else {
             throw new IllegalStateException(PATH_HOME_SETTING.getKey() + " is not configured");
         }
 
         if (PATH_CONF_SETTING.exists(settings)) {
-            configFile = PathUtils.get(fs, cleanPath(PATH_CONF_SETTING.get(settings)));
+            configFile = PathUtils.get(cleanPath(PATH_CONF_SETTING.get(settings)));
         } else {
             configFile = homeFile.resolve("config");
         }
 
         if (PATH_SCRIPTS_SETTING.exists(settings)) {
-            scriptsFile = PathUtils.get(fs, cleanPath(PATH_SCRIPTS_SETTING.get(settings)));
+            scriptsFile = PathUtils.get(cleanPath(PATH_SCRIPTS_SETTING.get(settings)));
         } else {
             scriptsFile = configFile.resolve("scripts");
         }
 
         if (PATH_PLUGINS_SETTING.exists(settings)) {
-            pluginsFile = PathUtils.get(fs, cleanPath(PATH_PLUGINS_SETTING.get(settings)));
+            pluginsFile = PathUtils.get(cleanPath(PATH_PLUGINS_SETTING.get(settings)));
         } else {
             pluginsFile = homeFile.resolve("plugins");
         }
@@ -144,7 +139,7 @@ public class Environment {
             dataFiles = new Path[dataPaths.size()];
             dataWithClusterFiles = new Path[dataPaths.size()];
             for (int i = 0; i < dataPaths.size(); i++) {
-                dataFiles[i] = PathUtils.get(fs, dataPaths.get(i));
+                dataFiles[i] = PathUtils.get(dataPaths.get(i));
                 dataWithClusterFiles[i] = dataFiles[i].resolve(ClusterName.clusterNameFromSettings(settings).value());
             }
         } else {
diff --git a/qa/evil-tests/src/test/java/org/elasticsearch/plugins/InstallPluginCommandTests.java b/qa/evil-tests/src/test/java/org/elasticsearch/plugins/InstallPluginCommandTests.java
index 7c51686..5b26102 100644
--- a/qa/evil-tests/src/test/java/org/elasticsearch/plugins/InstallPluginCommandTests.java
+++ b/qa/evil-tests/src/test/java/org/elasticsearch/plugins/InstallPluginCommandTests.java
@@ -27,10 +27,12 @@ import org.elasticsearch.Version;
 import org.elasticsearch.cli.MockTerminal;
 import org.elasticsearch.cli.UserError;
 import org.elasticsearch.common.io.PathUtils;
+import org.elasticsearch.common.io.PathUtilsForTesting;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.env.Environment;
 import org.elasticsearch.test.ESTestCase;
 import org.elasticsearch.test.PosixPermissionsResetter;
+import org.junit.After;
 
 import java.io.IOException;
 import java.io.InputStream;
@@ -71,12 +73,23 @@ public class InstallPluginCommandTests extends ESTestCase {
     private final FileSystem fs;
     private final boolean isPosix;
     private final boolean isReal;
+    private final String javaIoTmpdir;
 
     public InstallPluginCommandTests(FileSystem fs, Function<String, Path> temp) {
         this.fs = fs;
         this.temp = temp;
         this.isPosix = fs.supportedFileAttributeViews().contains("posix");
         this.isReal = fs == PathUtils.getDefaultFileSystem();
+        PathUtilsForTesting.installMock(fs);
+        javaIoTmpdir = System.getProperty("java.io.tmpdir");
+        System.setProperty("java.io.tmpdir", temp.apply("tmpdir").toString());
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        super.tearDown();
+        PathUtilsForTesting.teardown();
+        System.setProperty("java.io.tmpdir", javaIoTmpdir);
     }
 
     @ParametersFactory
@@ -124,7 +137,7 @@ public class InstallPluginCommandTests extends ESTestCase {
         Settings settings = Settings.builder()
             .put("path.home", home)
             .build();
-        return new Environment(fs, settings);
+        return new Environment(settings);
     }
 
     static Path createPluginDir(Function<String, Path> temp) throws IOException {
```